### PR TITLE
Support Add2App for iOS even if the main `Podfile` already has `post_install`

### PR DESCRIPTION
**Pull request by truongsinh**
## Description
Right now, https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps breaks if iOS project already has `post_install` snippet, see https://github.com/flutter/flutter/issues/26212
 
## Related Issues

- https://github.com/flutter/flutter/issues/24342#issuecomment-451926249
- https://stackoverflow.com/questions/40508957/invalid-podfile-file-specifying-multiple-post-install-hooks-is-unsupported
- https://github.com/CocoaPods/CocoaPods/issues/6172

## Tests

There is no test at the moment.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*

In https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps, the instruction should be changed to:

```
  flutter_application_path = 'path/to/my_flutter/'
  eval(File.read(File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')), binding)
# ...
    post_install do |installer|
      flutter_post_install(installer)
    end
```

- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by LinusU** [comment]
Yes, this is great! 🚀 

Since I have a deployment target of iOS 12 in my main app, my app doesn't build arm 32-bit code, but since some of the pods have a much lower deployment target they do. This leaves me with linking errors when I'm trying to archive the app, and the only fix I've found is to bump the deployment target to `12` on all my pods as well.

This leaves me in a state where I have to manually update the `podhelper.rb` every time I'm doing a release (since it's an autogenerated file I lose the changes all the time), to let the post-install hook add the updated deployment target...

ref: https://github.com/CocoaPods/CocoaPods/issues/4859#issuecomment-500838635

**Comment by blasten** [comment]
cc @jmagman

**Review by jmagman — CHANGES_REQUESTED** [review]
Doesn't this break CocoaPods in the module?  Now they don't have a post_install step.
What if we keep this flutter_post_install method, and add this to the end of the file:
```
post_install do |installer|
  flutter_post_install(installer)
 end
```

And then we can update the documentation to either require the module podhelper and call that method in an existing post_install, or do what mostly works now for host apps without CoocaPods and eval?

So the host app becomes:
```
require 'flutterapp/.ios/Flutter/podhelper.rb
post_install do |installer|
... existing post_install...
  flutter_post_install(installer)
end
```
Or stays:
```
eval(File.read(File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')), binding)
```

**Review by jmagman — CHANGES_REQUESTED** [review]
I think this is a good change.
1. The Podfile template needs to call the correct new method, otherwise the module will never install the pods.
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl
2. I _believe_ the framework_dir is the same directory as the podhelper `__dir__`, so I don't think it needs to be global or even passed in.
```
engine_dir = File.join(__dir__, 'engine')
```

File: packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl [hunk 1]
```diff
@@ -3,6 +3,9 @@ platform :ios, '8.0'
 target 'Runner' do
   flutter_application_path = '../'
```
Comment (jmagman) [inline packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl]: This doesn't seem like it does anything with the new global `$flutter_application_path`.

**Review by jmagman — CHANGES_REQUESTED** [review]

**Comment by jmagman** [comment]
You had a good solution to adding a method to the podhelper so existing Podfiles don't get clobbered.  I took a stab at a larger change that should hopefully solve things in a more complete, more CocoaPods-friendly way, and included the spirit of your change:
https://github.com/flutter/flutter/pull/36793
When that lands, you should be able to just update your Podfile targets that need the Flutter symbols, plugins, etc:
```
load File.join('my_flutter', '.ios', 'Flutter', 'podhelper.rb')
target 'MyApp' do
  install_all_flutter_pods 'my_flutter'
end
```
There will no longer be any post_install hooks required by Flutter.

**Comment by jmagman** [comment]
@truongsinh I'm hoping your issue is fixed by https://github.com/flutter/flutter/pull/36793.  You can migrate your Podfile as I suggested above, and try it on the master channel.  https://github.com/flutter/flutter/wiki/Flutter-build-release-channels
If it's not resolved, please file an issue and @ me!

**Comment by truongsinh** [comment]
Screen Shot 2019-07-31 at 1059 08

A quick try seems that the current master of Flutter assumes `FLUTTER_ROOT` is set, but it is not necessarily true, nor mentioned in https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#ios

**Comment by jmagman** [comment]
> Screen Shot 2019-07-31 at 1059 08
> 
> A quick try seems that the current master of Flutter assumes `FLUTTER_ROOT` is set, but it is not necessarily true, nor mentioned in https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#ios

You need to remove the xcode_backend build script phase you previously added to the app.  You no longer need it.  I was going to write up migration instructions tomorrow but you were too fast!

**Comment by truongsinh** [comment]
after removing `xcode_backend build scrip` and some other small changes, it works 🎉

**Comment by jmagman** [comment]
@truongsinh What were the "other small changes"? I want to make sure I cover everything in the migration instructions.

**Comment by truongsinh** [comment]
> @truongsinh What were the "other small changes"? I want to make sure I cover everything in the migration instructions.

It might be specific to our app only, but anyway 2 things

## 1

our app delegate extends/implements FlutterAppDelegate, FlutterStreamHandler

```swift
class AppDelegate: FlutterAppDelegate, FlutterStreamHandler {
```

down there in this class, we override
```swift
    override func applicationDidBecomeActive(_ application: UIApplication) {
        // .. do own own stuff

        // can no longer call supper (used to be able to call in previous Add2App version) because
        // `reason: '-[myApp.AppDelegate applicationDidBecomeActive:]: unrecognized selector sent to instance 0x6000011faf00'`
        // super.applicationDidBecomeActive(application)
        
    }
```

## 2
`flutterEngine` is no longer a `binaryMessenger` protocol, but need to change to `flutterEngine.binaryMessenger`

```diff
-        eventChannel = FlutterEventChannel.init(name: eventChannelName, binaryMessenger: flutterEngine)
-        methodChannel = FlutterMethodChannel.init(name: methodChannelName, binaryMessenger: flutterEngine)
+        eventChannel = FlutterEventChannel.init(name: eventChannelName, binaryMessenger: flutterEngine.binaryMessenger)
+        methodChannel = FlutterMethodChannel.init(name: methodChannelName, binaryMessenger: flutterEngine.binaryMessenger)
```

**Pull request closed without merging**